Added CI/CD statistics plugin

grantila opened the pull request:
## CI/CD statistics

This PR adds a new plugin to be used on the entity page. From the README:

This plugin shows charts of CI/CD pipeline durations over time. It expects to be used on the Software Catalog entity page, as it uses `useEntity` to figure out what component to get the build information for.

To use this plugin, you need to implement an API `CicdStatisticsApi` and bind it to the `cicdStatisticsApiRef`. This API is defined in `src/apis/types.ts` and is an interface with two functions, `getConfiguration()` and `fetchBuilds(options)`. This plugin will call `getConfiguration` to allow the implementation to specify defaults an settings for the UI. First time the UI shows, and each time the user changes filters and clicks `Update` to refresh the data, `fetchBuilds` is invoked with the filter options. The API implementation is the expected to fetch build information from somewhere, format it into a generic and rather simpe type `Build` (also defined in `types.ts`). The API can optionally signal completion for a progress bar in the UI.

When this plugin has fetched the builds, it will transpose the list of builds (and build stages) into a tree of build stages. As build pipelines sometimes change, certain stages might end or begin within the timerange of the view.

<img width="1495" alt="Screenshot 2022-01-13 at 11 12 53" src="https://user-images.githubusercontent.com/5362579/149334701-a5842dda-c7da-4c16-a785-bf8453e04752.png">

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+  // (undocumented)
+  fetchBuilds(options: FetchBuildsOptions): Promise<CicdState>;
+  // (undocumented)
+  getConfiguration(): Promise<Partial<CicdConfiguration>>;
```
Should we provide the `entity` as an option to this one as well? So that you can have different config for different CI/CD systems. Unless the config really is global, in which case it may be better to move to props on the extension instead.

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+  // (undocumented)
+  entity: Entity;
+  // (undocumented)
+  filterStatus: Array<FilterStatusType<'all'>>;
```
Think `FilterStatusType | 'all'` would be a nicer pattern? I had to jump to the type to figure out what it was doing

grantila replied:
Yes this is a bit iffy agreed.

Rugvip commented on plugins/cicd-statistics/src/plugin.ts:
```diff
@@ -0,0 +1,40 @@
+  createRoutableExtension({
+    component: () => import('./entity-page').then(m => m.EntityPageCicdCharts),
+    mountPoint: rootCatalogCicdStatsRouteRef,
+    name: 'cicd-statistics-page',
```
Should be `'EntityCicdStatisticsContent'`

grantila replied:
ok

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "rootCatalogCicdStatsRouteRef" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public (undocumented)
+export const rootCatalogCicdStatsRouteRef: RouteRef<undefined>;
```
Should actually avoid exporting this, it's accessible via `cicdStatisticsPlugin.routes.entityContent`. We'll figure out how to lint that at some point ._.

grantila replied:
ok sure

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+  {}
+>;
+
+// Warning: (ae-forgotten-export) The symbol "EntityPageCicdCharts" needs to be exported by the entry point index.d.ts
```
Would be nice to clean up all the warnings here, but I particularly recommend fixing the missing export warning, as the API Report does not show the complete API if you have these

grantila replied:
Hidden in an ocean of `ae-missing-release-tag`. Do we need these?

Rugvip replied:
Yeah if you wanna go through and add release tags that doesn't hurt. I'd call it optional at this point but it definitely cleans up the API report.

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "CicdConfiguration" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export interface CicdConfiguration {
```
Don't have any suggestion here but `Cicd` looks a bit awkward as part of these names xP
In case there's some other term that fits better

grantila replied:
Heh yeah, I'll let this linger for input, I'm open for suggestions!

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "BuildWithRaw" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export type BuildWithRaw<T = any> = Build & {
```
Do you think it would make sense to drop this and go for `Build<TRaw = unknown>` instead?

grantila replied:
So, we could actually remove this, and let implementors use something like it themselves. It's very useful though when constructing a list of these Builds and referring to the internal blobs you probably have in your particular implementation.

However, you shouldn't _have_ to use it (to e.g. set it to `undefined` or some other dummy), but if you **do** use it, you don't want it to be optional (`?`) so that you always need to `!` it. 🤔

So, **if** you want to use it in your API implementation, you can use `BuildWithRaw` with your own `MyBuildTypeThing` and return that back to this plugin, since it's compatible with `Build`. That was kind of the thinking here, although perhaps not very clear.

Rugvip commented on plugins/cicd-statistics/src/entity-page.tsx:
```diff
@@ -0,0 +1,175 @@
+  useEffect(() => {
+    if (
+      !chartableStagesState.error ||
+      chartableStagesState.error instanceof AbortError
```
I'd recommend `error.name === 'AbortError'` as `instanceof` is pretty unreliable when you start mixing package versions. It also makes the checking a bit more lax so that you end up supporting the error thrown by `fetch(..., { signal })`

grantila replied:
Good point!

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "AbortError" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export class AbortError extends Error {}
```
Bit unfortunate to export this here, but perhaps fine for now. Long-term I think we'd want this in `@backstage/errors` or something like that.

grantila replied:
That would make sense eventually

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+//
+// @public
+export interface Build {
+  buildType: FilterBranchType;
```
Confused about the naming here, is it a built or a branch type? :grin:

grantila replied:
Yes it is 😅
Will change

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "branchTypes" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public (undocumented)
+export const branchTypes: Array<FilterBranchType>;
```
Arguably fine to not export this? Or be you have some usage planned :grin:

grantila replied:
It's silly perhaps, but I was thinking that the *api implementation* might want it... It's not at all necessary though, I'll remove it.

Rugvip commented on plugins/cicd-statistics/api-report.md:
```diff
@@ -0,0 +1,182 @@
+// Warning: (ae-missing-release-tag) "Stage" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export interface Stage {
```
Think it would be worth calling this `BuildStage` for a bit of extra context? could see this being just fine too tho

grantila replied:
Yeah I kind of thought so too. It might be more clear. A Stage can have sub stages though, which are also of type Stage. BuildStage *does* (to me) sounds like the top-level stages.

I'm fine with either, happy to change if it makes more sense.

Rugvip reviewed (commented):
Awesome stuff :grin: 🎉 

Couple of thoughts on the API and small fixes

adamdmharvey commented on plugins/cicd-statistics/src/charts/status-chart.tsx:
```diff
@@ -0,0 +1,133 @@
+  Typography,
+} from '@material-ui/core';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+import { formatISO9075, parseISO } from 'date-fns';
```
I recall there was discussion on `date-fns` when date libraries were being decided upon, and the result was this ADRs:

https://backstage.io/docs/architecture-decisions/adrs-adr010
https://backstage.io/docs/architecture-decisions/adrs-adr012

Not suggesting to prevent a merge, per se, but this might need to be refactored later if it can't be moved to Luxon now?

grantila replied:
I've missed those ADR's, this makes total sense! I'll change to using Luxon and standardised formatting.

adamdmharvey commented on plugins/cicd-statistics/src/charts/utils.tsx:
```diff
@@ -0,0 +1,87 @@
+export function labelFormatterWithoutTime(epoch: number) {
+  return (
+    <span style={infoText}>
+      {formatISO9075(new Date(epoch), { representation: 'date' })}
```
Specifically this ADR https://backstage.io/docs/architecture-decisions/adrs-adr012 perhaps?

adamdmharvey reviewed (commented):
The charts are great. 👍🏻 Love them!

Apologies as I'm not quite understanding as a potential user - this supports ANY CI/CD tool?  Is there a backend somewhere with an assumption that the GitHub Actions or CircleCI or Jenkins plugins are storing results for this to read? Or is this one vendor specific for the moment? (a quick look through didn't show me exactly where the hooks are hitting)

Also one query re: the date ADRs.

awanlin commented:
@grantila this is really cool, great work! As one of the contributors to the Azure DevOps plugin would it be possible to coordinate some documentation around how to use this with Azure DevOps? Feel like it would make sense to do that with all the other CI/CD plugin contributors as they will get questions sent their way about this. Maybe we open a new issues for this, what do you think?

Rugvip commented:
@awanlin A possibility especially for quite generic plugins with lots of integrations like this would be to create `*-module-*` packages as hinted at in the [package architecture docs](https://backstage.io/docs/overview/architecture-overview#plugin-packages) and described in the [plugin structure ADR](https://backstage.io/docs/architecture-decisions/adrs-adr011). It could be a good pattern if we end up with a bit of a larger integration that might need a few azure dependencies and whatnot. Would be something like a `@backstage/plugin-cicd-statistics-module-azure` package similar to [@backstage/plugin-analytics-module-ga](https://github.com/backstage/backstage/tree/master/plugins/analytics-module-ga).

grantila commented:
> The charts are great. 👍🏻 Love them!
Thanks!

> Apologies as I'm not quite understanding as a potential user - this supports ANY CI/CD tool? Is there a backend somewhere with an assumption that the GitHub Actions or CircleCI or Jenkins plugins are storing results for this to read? Or is this one vendor specific for the moment? (a quick look through didn't show me exactly where the hooks are hitting)

So, this is not very vendor specific, or at least, the hope is for this to be generic enough for most systems out there. However it's built against one internal system only for now, so unfortunately, this plugin is in and of itself right now not useful to anyone without writing an API implementation to a CI/CD system. Many CI/CD systems don't natively have stored (and "quickly" retrievable) statistics of build time in build stages over time. The UI therefore has a progress bar to show the fetching progress as builds are retrieved.

There is no backend for this for now, that's a future interesting option. If so, that backend could store/cache data and have bindings/modules to CI/CD systems to fetch from.

The way this is used now, is by fetching all build time data straight from the browser to an internal CI/CD system. I know that GitHub Actions do store the builds way back, and [retrieving that data is possible](https://docs.github.com/en/rest/reference/actions), probably need to fetch artifact by artifact, which is how I've currently implemented this too. That backend idea sounds very interesting!

@awanlin @Rugvip yeah, having a package (`*-module-*`) per integration sounds reasonable. The fetching may be very slow though (which should be ok as an initial implementation), it's not a view you reload unless you have too. Perhaps it may make sense to actually think about a backend possibility for this, at least as one of the options. So, if a CI/CD has these statistics quickly retrievable, it could be a **code module** to this plugin, but for systems where one need to fetch entire build logs and do a lot of heavy parsing, going through a caching backend may be the way...

adamdmharvey commented:
Thanks for the useful content!  And I like the whole discussion around the ADR/module naming.

Given your context there, I'd suggest this likely shouldn't be approved to merge in without this context being added to the REDME. Otherwise, users of backstage will have no appreciation for how to use this?  So the README should say some f the points you've noted, like this won't work yet without a backlog (thus dont use :) ) and future work in this area, etc. That way others who are interested might contribute, but the average company can't use this plugin until there's a bit more work added to it?

Thanks for this contrib!  We support multiple CI tools at the moment (though trying to standardize) and consistently across tool chains like this could be really valuable.

grantila commented:
* Added per-stage status. In a build that has failed, e.g. the "install" stage might have succeeded. This allows for easier drill-down and overview of which stages fails more or less than others, and when.
 * Added count and duration inline in the same chart. Defaults to duration of successful builds and count for the rest. Duration for non-successful builds isn't often useful.
 * Added `med` median to each stage, which is probably more useful than average.
 * Added custom hide/collapse settings to hide (default) collapse stages with a peak (max) duration under a certain threshold.

<img width="1494" alt="Skärmavbild 2022-01-21 kl  17 59 02" src="https://user-images.githubusercontent.com/5362579/150569331-7bf4829a-c6e8-4162-928d-a08d1c46739a.png">

jhaals reviewed (commented):
yarn.lock is sad but otherwise let's get this :shipit:

grantila commented on plugins/cicd-statistics/src/hooks/use-cicd-statistics-api.ts:
```diff
@@ -0,0 +1,27 @@
+
+import { cicdStatisticsApiRef } from '../apis';
+
+export function useCicdStatisticsApi() {
```
IIRC, I thought the default was a bit harsh, and I had an idea of a fake dataset to display instead, which never happened though. Perhaps I should spend a few minutes constructing such, so that people can see the UI although without real data, to understand what the plugin does. WDYT?

freben replied:
As an aside, if you want to optionally get an an API instance, the canonical way is to do `useApiHolder().get(apiRef)`.

grantila commented:
> yarn.lock is sad but otherwise let's get this :shipit:

Yay! Let me fix the yarn.lock, but also add a small feature I'm about to finish now, namely charting manual re-triggering or builds. Most build systems have some type of "trigger reason" (e.g. from github webhook, or manual). Charting that is valuable, as it probably means people e.g. try to get through flaky tests.

grantila commented:
Force pushed many commits due to yarn.lock, but only the last three are added.

Added "trigger reason" to the API (optional) and UI. Most build systems have some property on builds being the source of the trigger, if it's manual or from an SCM or something else (internally scheduled e.g.)

The build state count (top chart) will now have an overlay (or, underlay rather) of the percentage of manual vs non-manually triggered builds. When the manually triggered builds are many (as the last couple of days in the screenshot), this indicates users need to re-trigger builds to pass the build pipeline - likely due to flaky tests or some other environment issue.

Also added zooming. Since it can take a long time to fetch all the build information, depending on the backend, it's handy to be able to zoom in the chart (and reset the zoom) without having to re-fetch all the build information. This is done by dragging an area in the chart. It will set this zoom for all charts, and you reset the zoom by just clicking in the chart.

<img width="1622" alt="Screenshot 2022-02-02 kl  16 22 30" src="https://user-images.githubusercontent.com/5362579/152194879-e9593d1c-b495-4960-8dec-9bdd8ec0c26a.png">

grantila commented:
@jhaals I'm pretty happy with the current state as a first version, IMO this can be merged.

freben commented on yarn.lock:
```diff
@@ -20804,6 +20988,26 @@ recharts@^1.8.5:
     recharts-scale "^0.4.2"
     reduce-css-calc "^1.3.0"
 
+recharts@^2.1.5:
```
Just a note here. The entire rest of the project uses `^1.8.5`, so this pulls in an extra dependency. It's not a huge deal in a lazy-loading world, but it would be nice to stay on one version if possible. Do you know if you could function fine with that older version and then we bump the entire repo at a later time to the 2.x line? If not, then fine to leave this as is

grantila replied:
1.8.5 is 2.5 years old. I wonder if we could bump them, I can give it a shot, unsure how to test them though apart from them simply compiling 🤔

freben replied:
As long as we don't do the bump as part of this PR. Let's keep things constrained

benjdlambert commented:
@grantila the yarn.lock has the wrong registry in there it's why the build is failing.

grantila commented:
Made progress bar loading more informative, as loading can take a while, and happen in multiple steps.

<img width="1403" alt="Skärmavbild 2022-02-07 kl  13 01 01" src="https://user-images.githubusercontent.com/5362579/152972390-f87a402a-1fed-43b4-a253-fa74618efcdc.png">

djamaile commented:
@grantila can't wait to somehow make it work with Gitlab, thanks for your work 😄
